app/models/deploy_metrics.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+# Used to collect metrics of cycle time(elapsed time needed to deliver a project to production)
+# Check out https://about.gitlab.com/2019/03/04/reduce-cycle-time/ for more details on cycle time
+# The main objective for this is to gain more insight on the efficiency of the end-to-end project development process
+# There are 2 key metrics:
+# 1. PR - Production: this will show the time needed for a pull request to get deployed to production environment
+# 2. Staging - Production: this will show the time needed to deliver from staging(non-production environment)
+#                          to production environment.
+class DeployMetrics
+  def initialize(deploy)
+    @deploy = deploy
+  end
+
+  def cycle_time
+    if @deploy&.production && @deploy&.status == "succeeded"
+      {pr_production: pr_production, staging_production: staging_production}
+    else
+      {pr_production: 0, staging_production: 0}
+    end
+  end
+
+  private
+
+  # Calculated by averaging deployment end time minus pull requests' creation time
+  def pr_production
+    prs = @deploy&.changeset&.pull_requests || []
+    prs.any? ? prs.sum { |pr| @deploy.updated_at.to_i - pr.created_at.to_i } / prs.size : 0
+  end
+
+  # Calculated by taking production deployment end time minus the very first staging deployment end time
+  def staging_production
+    jobs = Job.where(commit: @deploy&.commit).pluck(:id)
+    stages = Stage.where(project: @deploy&.project)
+    stages = stages.reject(&:production?)
+
+    first_staging_deploy = Deploy.where(job: jobs, stage: stages).reorder(:id).first
+    first_staging_deploy ? @deploy.updated_at.to_i - first_staging_deploy.updated_at.to_i : 0
```

Comment:
we can only count the first production deploy, so need to do the same logic as the first_staging_deploy but `first_production_deploy `

Reply:
alright, agree